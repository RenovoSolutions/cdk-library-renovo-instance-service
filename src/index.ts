import * as ec2 from '@aws-cdk/aws-ec2';
import * as iam from '@aws-cdk/aws-iam';
import * as cdk from '@aws-cdk/core';
import { ManagedInstanceRole } from '@renovosolutions/aws-cdk-managed-instance-role';

export interface IInstanceServiceProps {
  /**
   * The Amazon Machine Image (AMI) to launch the target instance with
   */
  ami: ec2.IMachineImage;
  /**
   * Whether or not to enable logging to Cloudwatch Logs
   *
   *
   * @default true
   */
  enableCloudwatchLogs?: boolean;
  // Add domain join
  // Add SSM
}

export class InstanceService extends cdk.Construct {

  // public readonly instance: ec2.Instance

  constructor(scope: cdk.Construct, id: string, props: IInstanceServiceProps) {
    super(scope, id);

    let managedPolicies = [];

    // Logging configuration
    props.enableCloudwatchLogs = (props.enableCloudwatchLogs === undefined) ? true : props.enableCloudwatchLogs;

    if (props.enableCloudwatchLogs) {
      let logResources:string[] = [];
      let os:string = ec2.OperatingSystemType[props.ami.getImage(this).osType].toLowerCase();
      console.log(os);
      const logGroups:string[] = [
        `/${ os }/logs`,
        `/${ os }/logs/*`,
      ];
      logGroups.forEach( (g) => {
        logResources.push(`arn:aws:logs:${ cdk.Stack.of(this).region }:${ cdk.Stack.of(this).account }:log-group:${ g }`);
      });

      managedPolicies.push(new iam.ManagedPolicy(this, 'loggingPolicy', {
        description: 'Allow instance to log system logs to Cloudwatch',
        statements: [
          new iam.PolicyStatement({
            effect: iam.Effect.ALLOW,
            actions: [
              'logs:PutLogEvents',
              'logs:CreateLogGroup',
              'logs:CreateLogStream',
            ],
            resources: logResources,
          }),
        ],
      }));
    }

    new ManagedInstanceRole(this, 'instanceRole', {
      domainJoinEnabled: false,
      ssmManagementEnabled: true,
      managedPolicies,
    });
  }
}
