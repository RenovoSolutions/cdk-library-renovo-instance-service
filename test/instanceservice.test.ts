import { expect as expectCDK, haveResourceLike, SynthUtils } from '@aws-cdk/assert';
import * as ec2 from '@aws-cdk/aws-ec2';
import * as cdk from '@aws-cdk/core';
import { ec2ImageToOsString, InstanceService, ManagedLoggingPolicy } from '../src/index';

test('Snapshot', () => {
  const app = new cdk.App();
  const stack = new cdk.Stack(app, 'TestStack', {
    env: {
      account: '123456789012', // not a real account
      region: 'us-east-1',
    },
  });

  const ami = ec2.MachineImage.lookup({
    name: 'Windows_Server-2022-English-Full-Base',
    windows: true,
  });

  new InstanceService(stack, 'instanceService', {
    ami,
  });

  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});

test('DefaultLinuxService', () => {
  const app = new cdk.App();
  const stack = new cdk.Stack(app, 'TestStack', {
    env: {
      account: '223456789012', // not a real account
      region: 'us-east-1',
    },
  });

  const amiLinux = ec2.MachineImage.lookup({
    name: 'ubuntu/images/hvm-ssd/ubuntu-focal-20.04-amd64-server*',
    owners: ['099720109477'],
    windows: false,
  });

  new InstanceService(stack, 'instanceService', {
    ami: amiLinux,
  });

  expectCDK(stack).to(haveResourceLike('AWS::IAM::ManagedPolicy', {
    PolicyDocument: {
      Statement: [{
        Resource: [
          'arn:aws:logs:us-east-1:223456789012:log-group:/linux/logs',
          'arn:aws:logs:us-east-1:223456789012:log-group:/linux/logs/*',
        ],
      }],
    },
  }));
});

test('DefaultWindowsService', () => {
  const app = new cdk.App();
  const stack = new cdk.Stack(app, 'TestStack', {
    env: {
      account: '123456789012', // not a real account
      region: 'us-east-1',
    },
  });

  const amiWin = ec2.MachineImage.lookup({
    name: 'Windows_Server-2022-English-Full-Base',
    windows: true,
  });

  new InstanceService(stack, 'instanceService', {
    ami: amiWin,
  });

  expectCDK(stack).to(haveResourceLike('AWS::IAM::ManagedPolicy', {
    PolicyDocument: {
      Statement: [{
        Resource: [
          'arn:aws:logs:us-east-1:123456789012:log-group:/windows/logs',
          'arn:aws:logs:us-east-1:123456789012:log-group:/windows/logs/*',
        ],
      }],
    },
  }));
});

test('Policies only', () => {
  const app = new cdk.App();
  const stack = new cdk.Stack(app, 'TestStack', {
    env: {
      account: '123456789012', // not a real account
      region: 'us-east-1',
    },
  });

  new ManagedLoggingPolicy(stack, 'policy', {
    os: 'windows',
  });

  const amiLinux = ec2.MachineImage.lookup({
    name: 'ubuntu/images/hvm-ssd/ubuntu-focal-20.04-amd64-server*',
    owners: ['099720109477'],
    windows: false,
  });

  new ManagedLoggingPolicy(stack, 'linuxpolicy', {
    os: ec2ImageToOsString(stack, amiLinux),
  });

  expectCDK(stack).to(haveResourceLike('AWS::IAM::ManagedPolicy', {
    PolicyDocument: {
      Statement: [{
        Resource: [
          'arn:aws:logs:us-east-1:123456789012:log-group:/windows/logs',
          'arn:aws:logs:us-east-1:123456789012:log-group:/windows/logs/*',
        ],
      }],
    },
  }));

  expectCDK(stack).to(haveResourceLike('AWS::IAM::ManagedPolicy', {
    PolicyDocument: {
      Statement: [{
        Resource: [
          'arn:aws:logs:us-east-1:123456789012:log-group:/linux/logs',
          'arn:aws:logs:us-east-1:123456789012:log-group:/linux/logs/*',
        ],
      }],
    },
  }));
});
