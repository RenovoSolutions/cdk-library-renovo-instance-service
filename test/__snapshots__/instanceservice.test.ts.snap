// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
Object {
  "Resources": Object {
    "instanceServiceinstanceRoleinstanceProfile95F55675": Object {
      "Properties": Object {
        "Roles": Array [
          Object {
            "Ref": "instanceServiceinstanceRoleroleC95CA18D",
          },
        ],
      },
      "Type": "AWS::IAM::InstanceProfile",
    },
    "instanceServiceinstanceRoleroleC95CA18D": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "ec2.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Ref": "instanceServiceloggingPolicy5CD2F985",
          },
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/AmazonSSMManagedInstanceCore",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "instanceServiceloggingPolicy5CD2F985": Object {
      "Properties": Object {
        "Description": "Allow instance to log system logs to Cloudwatch",
        "Path": "/",
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "logs:PutLogEvents",
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
              ],
              "Effect": "Allow",
              "Resource": Array [
                "arn:aws:logs:us-east-1:123456789012:log-group:/windows/logs",
                "arn:aws:logs:us-east-1:123456789012:log-group:/windows/logs/*",
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::ManagedPolicy",
    },
  },
}
`;
